sklearn/metrics/pairwise.py

```diff
@@ -1985,16 +2002,29 @@ def _pairwise_callable(X, Y, metric, ensure_all_finite=True, **kwds):
         # No input dimension checking done for custom metrics (left to user)
         ensure_2d=False,
     )
+    _, _, dtype_float = _find_floating_dtype_allow_sparse(X, Y, xp=xp)
+
+    def _get_slice(array, index):
+        # TODO : Support for 1D shapes in scipy sparse arrays (COO, DOK and CSR
+        # formats) only added in 1.14. We must return 2D array until min scipy 1.14
+        # i.e. below 2 lines can be removed once min scipy >= 1.14
+        if issparse(array):
+            return array[[index], :]
+        # When `metric` is a callable, 1D input arrays allowed, in which case
+        # scalar should be returned.
+        if array.ndim == 1:
+            return array[index]
+        else:
+            return array[index, ...]
 
     if X is Y:
         # Only calculate metric for upper triangle
-        out = np.zeros((X.shape[0], Y.shape[0]), dtype="float")
+        out = xp.zeros((X.shape[0], Y.shape[0]), dtype=dtype_float)
         iterator = itertools.combinations(range(X.shape[0]), 2)
         for i, j in iterator:
-            # scipy has not yet implemented 1D sparse slices; once implemented this can
-            # be removed and `arr[ind]` can be simply used.
-            x = X[[i], :] if issparse(X) else X[i]
-            y = Y[[j], :] if issparse(Y) else Y[j]
+            x = _get_slice(X, i)
+            # y = _get_slice(Y, j)
+            y = Y[j, ...] if Y.ndim == 2 else Y[j]
```

Comment:
Why not call `y = _get_slice(Y, j)` here?

Reply:
I forgot to uncomment, I was testing in different scenarios :grimacing: